File: src/credit_workflow/planner.py
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass

from .models import ActivityBundle, DemandRecord


@dataclass(frozen=True, slots=True)
class BundleCandidate:
    bundle: ActivityBundle
    sign_up_id: str
    user_id: str


@dataclass(frozen=True, slots=True)
class PlannableDemand:
    demand: DemandRecord
    candidates: tuple[BundleCandidate, ...]


@dataclass(frozen=True, slots=True)
class DemandAllocation:
    demand: DemandRecord
    assignments: tuple[BundleCandidate, ...]
    planned_value_cent: int
    delta_cent: int


@dataclass(frozen=True, slots=True)
class PlanningResult:
    allocations: tuple[DemandAllocation, ...]
    bundle_assignment_counts: dict[str, int]


def bundle_key(bundle: ActivityBundle) -> str:
    credit_ids = ",".join(sorted(item.credit_id for item in bundle.credit_items))
    return f"{bundle.activity_id}:{bundle.credit_type}:{credit_ids}"


def _candidate_balance_key(
    assignments: tuple[BundleCandidate, ...],
    usage_counts: dict[str, int],
) -> tuple[float, float, int, tuple[str, ...]]:
    post_usage_rates = []
    keys = []
    for assignment in assignments:
        key = bundle_key(assignment.bundle)
        keys.append(key)
        post_usage_rates.append(
            (usage_counts.get(key, 0) + 1) / assignment.bundle.bundle_capacity
        )

    return (
        sum(rate * rate for rate in post_usage_rates),
        max(post_usage_rates, default=0.0),
        len(assignments),
        tuple(sorted(keys)),
    )


def _choose_best_assignments(
    target_value_cent: int,
    candidates: tuple[BundleCandidate, ...],
    usage_counts: dict[str, int],
    *,
    prefer_overflow: bool,
) -> tuple[BundleCandidate, ...]:
    if not candidates:
        return ()

    max_bundle_value = max(candidate.bundle.bundle_value_cent for candidate in candidates)
    total_limit = target_value_cent + max_bundle_value

    states: dict[int, tuple[BundleCandidate, ...]] = {0: ()}
    for candidate in candidates:
        next_states = dict(states)
        for total, assignments in sorted(states.items(), reverse=True):
            next_total = total + candidate.bundle.bundle_value_cent
            if next_total > total_limit:
                continue

            next_assignments = assignments + (candidate,)
            existing = next_states.get(next_total)
            if existing is None or _candidate_balance_key(
                next_assignments, usage_counts
            ) < _candidate_balance_key(existing, usage_counts):
                next_states[next_total] = next_assignments

        states = next_states

    totals = [total for total in states if total > 0]
    if not totals:
        return ()

    exact_totals = [total for total in totals if total == target_value_cent]
    if exact_totals:
        exact_total = exact_totals[0]
        return states[exact_total]

    if prefer_overflow:
        overflow_totals = [total for total in totals if total > target_value_cent]
        if overflow_totals:
            best_total = min(overflow_totals)
            return states[best_total]
    else:
        shortage_totals = [total for total in totals if total < target_value_cent]
        if shortage_totals:
            best_total = max(shortage_totals)
            return states[best_total]
        return ()

    best_total = max(totals)
    return states[best_total]


def plan_student_allocation(
    demand: DemandRecord,
    candidates: tuple[BundleCandidate, ...],
    usage_counts: dict[str, int],
    *,
    prefer_overflow: bool = True,
) -> DemandAllocation:
    available_candidates = tuple(
        candidate
        for candidate in candidates
        if usage_counts.get(bundle_key(candidate.bundle), 0) < candidate.bundle.bundle_capacity
    )
    assignments = _choose_best_assignments(
        demand.requested_value_cent,
        available_candidates,
        usage_counts,
        prefer_overflow=prefer_overflow,
    )
    planned_value_cent = sum(
        assignment.bundle.bundle_value_cent for assignment in assignments
    )

    return DemandAllocation(
        demand=demand,
        assignments=assignments,
        planned_value_cent=planned_value_cent,
        delta_cent=planned_value_cent - demand.requested_value_cent,
    )


class Planner:
    def __init__(self, bundle_assignment_counts: dict[str, int] | None = None):
        self.bundle_assignment_counts = dict(bundle_assignment_counts or {})

    def _sort_key(
        self,
        plannable_demand: PlannableDemand,
        original_index: int,
    ) -> tuple[int, int, int, int]:
        available_candidates = [
            candidate
            for candidate in plannable_demand.candidates
            if self.bundle_assignment_counts.get(bundle_key(candidate.bundle), 0)
            < candidate.bundle.bundle_capacity
        ]
        reachable_total = sum(
            candidate.bundle.bundle_value_cent for candidate in available_candidates
        )
        return (
            len(available_candidates),
            reachable_total,
            -plannable_demand.demand.requested_value_cent,
            original_index,
        )

    @staticmethod
    def _apply_assignment_counts_delta(
        assignment_counts: dict[str, int],
        assignments: tuple[BundleCandidate, ...],
        delta: int,
    ) -> None:
        for assignment in assignments:
            key = bundle_key(assignment.bundle)
            assignment_counts[key] = assignment_counts.get(key, 0) + delta

    @staticmethod
    def _incremental_upgrade_cost(
        current_assignments: tuple[BundleCandidate, ...],
        upgraded_assignments: tuple[BundleCandidate, ...],
    ) -> int:
        current_counts: dict[str, int] = defaultdict(int)
        upgraded_counts: dict[str, int] = defaultdict(int)
        for assignment in current_assignments:
            current_counts[bundle_key(assignment.bundle)] += 1
        for assignment in upgraded_assignments:
            upgraded_counts[bundle_key(assignment.bundle)] += 1

        all_keys = set(current_counts) | set(upgraded_counts)
        return sum(
            max(0, upgraded_counts.get(key, 0) - current_counts.get(key, 0))
            for key in all_keys
        )

    def _upgrade_shortages_with_remaining_capacity(
        self,
        plannable_demands: tuple[PlannableDemand, ...],
        ordered_indices: list[int],
        allocations_by_index: dict[int, DemandAllocation],
        assignment_counts: defaultdict[str, int],
    ) -> None:
        pending = [
            index
            for index in ordered_indices
            if allocations_by_index[index].delta_cent < 0
        ]

        while pending:
            upgrade_options: list[
                tuple[
                    tuple[object, ...],
                    int,
                    DemandAllocation,
                    tuple[BundleCandidate, ...],
                ]
            ] = []

            for index in pending:
                allocation = allocations_by_index[index]
                if allocation.delta_cent >= 0:
                    continue

                demand = plannable_demands[index].demand
                current_assignments = allocation.assignments
                temp_counts = defaultdict(int, assignment_counts)
                self._apply_assignment_counts_delta(
                    temp_counts,
                    current_assignments,
                    delta=-1,
                )

                upgraded = plan_student_allocation(
                    demand,
                    plannable_demands[index].candidates,
                    temp_counts,
                    prefer_overflow=True,
                )
                if upgraded.planned_value_cent < demand.requested_value_cent:
                    continue

                if upgraded.assignments == current_assignments:
                    continue

                incremental_cost = self._incremental_upgrade_cost(
                    current_assignments,
                    upgraded.assignments,
                )
                option_key = (
                    incremental_cost,
                    upgraded.delta_cent,
                    len(upgraded.assignments),
                    _candidate_balance_key(upgraded.assignments, temp_counts),
                    index,
                )
                upgrade_options.append(
                    (
                        option_key,
                        index,
                        upgraded,
                        current_assignments,
                    )
                )

            if not upgrade_options:
                break

            _, selected_index, selected_upgrade, selected_current = min(
                upgrade_options, key=lambda item: item[0]
            )
            self._apply_assignment_counts_delta(
                assignment_counts,
                selected_current,
                delta=-1,
            )
            self._apply_assignment_counts_delta(
                assignment_counts,
                selected_upgrade.assignments,
                delta=1,
            )
            allocations_by_index[selected_index] = selected_upgrade
            pending = [
                index
                for index in pending
                if allocations_by_index[index].delta_cent < 0
            ]

    def plan(self, plannable_demands: tuple[PlannableDemand, ...]) -> PlanningResult:
        allocations_by_index: dict[int, DemandAllocation] = {}
        assignment_counts = defaultdict(int, self.bundle_assignment_counts)

        grouped_indices: dict[str, list[int]] = defaultdict(list)
        for index, plannable_demand in enumerate(plannable_demands):
            grouped_indices[plannable_demand.demand.credit_type].append(index)

        for credit_type in sorted(grouped_indices):
            ordered_indices = sorted(
                grouped_indices[credit_type],
                key=lambda index: self._sort_key(plannable_demands[index], index),
            )
            for index in ordered_indices:
                allocation = plan_student_allocation(
                    plannable_demands[index].demand,
                    plannable_demands[index].candidates,
                    assignment_counts,
                    prefer_overflow=False,
                )
                allocations_by_index[index] = allocation
                for assignment in allocation.assignments:
                    assignment_counts[bundle_key(assignment.bundle)] += 1

            self._upgrade_shortages_with_remaining_capacity(
                plannable_demands,
                ordered_indices,
                allocations_by_index,
                assignment_counts,
            )

        allocations = tuple(
            allocations_by_index[index] for index in range(len(plannable_demands))
        )
        return PlanningResult(
            allocations=allocations,
            bundle_assignment_counts=dict(assignment_counts),
        )

File: server/domain/planner.ts
import type { ActivityBundle, BundleCandidate, DemandAllocation, DemandRecord, PlannableDemand, PlanningResult } from "./models";

export function bundleKey(bundle: ActivityBundle): string {
  const creditIds = bundle.creditItems.map((item) => item.creditId).sort().join(",");
  return `${bundle.activityId}:${bundle.creditType}:${creditIds}`;
}

function candidateBalanceKey(assignments: BundleCandidate[], usageCounts: Record<string, number>): [number, number, number, string] {
  const rates: number[] = [];
  const keys: string[] = [];
  for (const assignment of assignments) {
    const key = bundleKey(assignment.bundle);
    keys.push(key);
    rates.push(((usageCounts[key] ?? 0) + 1) / Math.max(1, assignment.bundle.bundleCapacity));
  }
  return [
    rates.reduce((sum, rate) => sum + rate * rate, 0),
    Math.max(0, ...rates),
    assignments.length,
    keys.sort().join("|"),
  ];
}

function compareTuple(left: readonly (number | string)[], right: readonly (number | string)[]): number {
  for (let index = 0; index < left.length; index += 1) {
    if (left[index] < right[index]) {
      return -1;
    }
    if (left[index] > right[index]) {
      return 1;
    }
  }
  return 0;
}

function chooseBestAssignments(targetValueCent: number, candidates: BundleCandidate[], usageCounts: Record<string, number>, preferOverflow: boolean): BundleCandidate[] {
  if (candidates.length === 0) {
    return [];
  }
  const maxBundleValue = Math.max(...candidates.map((candidate) => candidate.bundle.bundleValueCent));
  const totalLimit = targetValueCent + maxBundleValue;
  let states = new Map<number, BundleCandidate[]>([[0, []]]);

  for (const candidate of candidates) {
    const nextStates = new Map(states);
    const entries = [...states.entries()].sort((left, right) => right[0] - left[0]);
    for (const [total, assignments] of entries) {
      const nextTotal = total + candidate.bundle.bundleValueCent;
      if (nextTotal > totalLimit) {
        continue;
      }
      const nextAssignments = [...assignments, candidate];
      const existing = nextStates.get(nextTotal);
      if (!existing || compareTuple(candidateBalanceKey(nextAssignments, usageCounts), candidateBalanceKey(existing, usageCounts)) < 0) {
        nextStates.set(nextTotal, nextAssignments);
      }
    }
    states = nextStates;
  }

  const totals = [...states.keys()].filter((total) => total > 0);
  if (totals.length === 0) {
    return [];
  }
  if (states.has(targetValueCent)) {
    return states.get(targetValueCent) ?? [];
  }
  if (preferOverflow) {
    const overflowTotals = totals.filter((total) => total > targetValueCent);
    if (overflowTotals.length > 0) {
      return states.get(Math.min(...overflowTotals)) ?? [];
    }
  } else {
    const shortageTotals = totals.filter((total) => total < targetValueCent);
    if (shortageTotals.length > 0) {
      return states.get(Math.max(...shortageTotals)) ?? [];
    }
    return [];
  }
  return states.get(Math.max(...totals)) ?? [];
}

export function planStudentAllocation(demand: DemandRecord, candidates: BundleCandidate[], usageCounts: Record<string, number>, preferOverflow = true): DemandAllocation {
  const availableCandidates = candidates.filter((candidate) => (usageCounts[bundleKey(candidate.bundle)] ?? 0) < candidate.bundle.bundleCapacity);
  const assignments = chooseBestAssignments(demand.requestedValueCent, availableCandidates, usageCounts, preferOverflow);
  const plannedValueCent = assignments.reduce((sum, assignment) => sum + assignment.bundle.bundleValueCent, 0);
  return {
    demand,
    assignments,
    plannedValueCent,
    deltaCent: plannedValueCent - demand.requestedValueCent,
    enabled: true,
  };
}

export class Planner {
  private readonly initialCounts: Record<string, number>;

  constructor(bundleAssignmentCounts: Record<string, number> = {}) {
    this.initialCounts = { ...bundleAssignmentCounts };
  }

  plan(plannableDemands: PlannableDemand[]): PlanningResult {
    const allocationsByIndex = new Map<number, DemandAllocation>();
    const assignmentCounts: Record<string, number> = { ...this.initialCounts };
    const grouped = new Map<string, number[]>();
    plannableDemands.forEach((plannableDemand, index) => {
      const values = grouped.get(plannableDemand.demand.creditType) ?? [];
      values.push(index);
      grouped.set(plannableDemand.demand.creditType, values);
    });

    for (const creditType of [...grouped.keys()].sort((left, right) => left.localeCompare(right, "zh-CN"))) {
      const orderedIndices = [...(grouped.get(creditType) ?? [])].sort((left, right) => {
        return compareTuple(this.sortKey(plannableDemands[left], left, assignmentCounts), this.sortKey(plannableDemands[right], right, assignmentCounts));
      });
      for (const index of orderedIndices) {
        const allocation = planStudentAllocation(plannableDemands[index].demand, plannableDemands[index].candidates, assignmentCounts, false);
        allocationsByIndex.set(index, allocation);
        applyAssignmentCountsDelta(assignmentCounts, allocation.assignments, 1);
      }
      this.upgradeShortages(plannableDemands, orderedIndices, allocationsByIndex, assignmentCounts);
    }

    return {
      allocations: plannableDemands.map((_, index) => allocationsByIndex.get(index) ?? zeroAllocation(plannableDemands[index].demand)),
      bundleAssignmentCounts: assignmentCounts,
    };
  }

  private sortKey(plannableDemand: PlannableDemand, originalIndex: number, assignmentCounts: Record<string, number>): [number, number, number, number] {
    const availableCandidates = plannableDemand.candidates.filter((candidate) => (assignmentCounts[bundleKey(candidate.bundle)] ?? 0) < candidate.bundle.bundleCapacity);
    const reachableTotal = availableCandidates.reduce((sum, candidate) => sum + candidate.bundle.bundleValueCent, 0);
    return [
      availableCandidates.length,
      reachableTotal,
      -plannableDemand.demand.requestedValueCent,
      originalIndex,
    ];
  }

  private upgradeShortages(plannableDemands: PlannableDemand[], orderedIndices: number[], allocationsByIndex: Map<number, DemandAllocation>, assignmentCounts: Record<string, number>): void {
    let pending = orderedIndices.filter((index) => (allocationsByIndex.get(index)?.deltaCent ?? 0) < 0);
    while (pending.length > 0) {
      const options: Array<{ key: (number | string)[]; index: number; upgraded: DemandAllocation; current: BundleCandidate[] }> = [];
      for (const index of pending) {
        const allocation = allocationsByIndex.get(index);
        if (!allocation || allocation.deltaCent >= 0) {
          continue;
        }
        const tempCounts = { ...assignmentCounts };
        applyAssignmentCountsDelta(tempCounts, allocation.assignments, -1);
        const upgraded = planStudentAllocation(plannableDemands[index].demand, plannableDemands[index].candidates, tempCounts, true);
        if (upgraded.plannedValueCent < plannableDemands[index].demand.requestedValueCent || sameAssignments(upgraded.assignments, allocation.assignments)) {
          continue;
        }
        options.push({
          key: [
            incrementalUpgradeCost(allocation.assignments, upgraded.assignments),
            upgraded.deltaCent,
            upgraded.assignments.length,
            ...candidateBalanceKey(upgraded.assignments, tempCounts),
            index,
          ],
          index,
          upgraded,
          current: allocation.assignments,
        });
      }
      if (options.length === 0) {
        break;
      }
      const selected = options.sort((left, right) => compareTuple(left.key, right.key))[0];
      applyAssignmentCountsDelta(assignmentCounts, selected.current, -1);
      applyAssignmentCountsDelta(assignmentCounts, selected.upgraded.assignments, 1);
      allocationsByIndex.set(selected.index, selected.upgraded);
      pending = orderedIndices.filter((index) => (allocationsByIndex.get(index)?.deltaCent ?? 0) < 0);
    }
  }
}

function zeroAllocation(demand: DemandRecord): DemandAllocation {
  return {
    demand,
    assignments: [],
    plannedValueCent: 0,
    deltaCent: -demand.requestedValueCent,
    enabled: true,
  };
}

function applyAssignmentCountsDelta(counts: Record<string, number>, assignments: BundleCandidate[], delta: number): void {
  for (const assignment of assignments) {
    const key = bundleKey(assignment.bundle);
    counts[key] = (counts[key] ?? 0) + delta;
  }
}

function sameAssignments(left: BundleCandidate[], right: BundleCandidate[]): boolean {
  return left.map((item) => bundleKey(item.bundle)).join("|") === right.map((item) => bundleKey(item.bundle)).join("|");
}

function incrementalUpgradeCost(currentAssignments: BundleCandidate[], upgradedAssignments: BundleCandidate[]): number {
  const currentCounts = countAssignments(currentAssignments);
  const upgradedCounts = countAssignments(upgradedAssignments);
  const keys = new Set([...Object.keys(currentCounts), ...Object.keys(upgradedCounts)]);
  let cost = 0;
  for (const key of keys) {
    cost += Math.max(0, (upgradedCounts[key] ?? 0) - (currentCounts[key] ?? 0));
  }
  return cost;
}

function countAssignments(assignments: BundleCandidate[]): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const assignment of assignments) {
    const key = bundleKey(assignment.bundle);
    counts[key] = (counts[key] ?? 0) + 1;
  }
  return counts;
}
